packages/plugins/basic-pipeline/src/index.js

```diff
@@ -28,6 +27,11 @@ class BasicPipeline {
     this.plugins = embark.pluginsAPI;
     this.pipelinePlugins = this.plugins.getPluginsFor('pipeline');
     this.pipelineConfig = embark.config.pipelineConfig;
+
+    const env = embark.config.env;
+    this.webpackConfigName = embark.pluginConfig[env] && embark.pluginConfig[env].webpackConfigName ?
+      embark.pluginConfig[env].webpackConfigName : env;
```

Comment:
Isn't `env` an arbitratry string? So, yes, Embark defaults to `development`, but using `embark build` doesn't use `env = 'production'`, also, the user could use `prod` instead if he wanted to.
Does that break the behaviour you,re trying to have here?

Reply:
Right now `embark build` defaults to `production` if no `--pipeline` option is specified. This is done here: https://github.com/embarklabs/embark/pull/2347/files#diff-b458cce19e6af687f43b8d768f741052L148

However, since we moved the pipeline into a optional plugin, that options has been ignored anyways. At this point it's actually not getting any webpackConfig and is therefore non-functioning.

>  also, the user could use prod instead if he wanted to.

True. This assumes that the users uses environments specified in other configs (where they can be aribitrary strings as well)

Reply:
So, I thought about it more, and though it makes sense that it defaults to `development` most of the time, as soon as someone uses a custom env, the pipeline will be broken. Eg, in Teller, we often use the `testnet` or `rinkeby` envs. Then, the pipeline would just crap the bed.

So, IMO, we should always set the pipeline to `development` mode and then use the plugin config you added to tell the pipeline which env should be in production. Eg:
```
 "plugins": {
    "embark-basic-pipeline": {
      "production": {
        webpackConfigName: "production",
        envs: ['mainnet']
      }
    }
  }
```
But at the same time, it's usually better to always `run` with development, because you can make adjustments easily and not have to wait forever.
So maybe something about the command would be better:
```
 "plugins": {
    "embark-basic-pipeline": {
      "production": {
        webpackConfigName: "production",
        commands: ['build', 'upload']
      }
    }
  }
```
You can get the command using the context: `this.embark.currentContext.includes('build');`.

Let me know what you think.

Reply:
Some good thoughts @jrainville. Firstly, I'd like to point out that this is not the first plugin that has needed environment-specific settings. At some point we need to address changes that handle environment-specific plugin configuration at a global level. At the very least, Embark should be able to apply a current environment to the `plugins` config to get the that plugin's current configuration. For example, given a config like:
```
"plugins": {
    "embark-basic-pipeline": {
      "production": {
        webpackConfigName: "production"
      },
      "development": {
        webpackConfigName: "development"
      }
    }
  }
```
If the current environment was `development`, then the value of `pluginConfig.webpackConfigName` would be `"development"`.

IMO, this could be written as part of this PR, or handled in a separate PR.

As to how the config should be organised, my opinion is that @PascalPrecht's original format is correct, however, as @jrainville pointed out, the default should always be `development`, and only then overridden by the config. So this would mean:

Assuming a plugin config of:
```
"plugins": {
    "embark-basic-pipeline": {
      "production": {
        webpackConfigName: "production"
      },
      "development": {
        webpackConfigName: "development"
      }
    }
  }
```
We could get the config name by:
```
const env = embark.config.env;
this.webpackConfigName = embark.pluginConfig[env]?.webpackConfigName ?? 'development';
```
If the environment doesn't exist in the plugin config, it defaults to development.

I don't really see the need to limit this by command either. If the user has a need to build with a production webpack config, they then do `embark build production`. This makes sense to me.

This config format (specifying config by environment) is how all the other configs work and should be familiar to the user. In addition, by keeping the config this way, we set ourselves up for allowing Embark to create environment-specific plugin configs as mentioned in the first part of my comment.

Reply:
> IMO, this could be written as part of this PR, or handled in a separate PR.

I might be missing something here, but this is actually exactly what this PR is doing.

> If the environment doesn't exist in the plugin config, it defaults to development.
>  Eg, in Teller, we often use the testnet or rinkeby envs. Then, the pipeline would just crap the bed.

@emizzle @jrainville good point, probably better to not require a plugin user to add every possible custom environment and just default to `development`. Will update.

Reply:
> I might be missing something here, but this is actually exactly what this PR is doing.
It is *for this plugin*. But what I'm suggesting is to allow environment-specific plugin config that gets applied automatically for *any* plugin.